nengo/synapses.py

```diff
@@ -466,8 +560,332 @@ class Alpha(LinearFilter):
+
+    .. testcode::
+
+        from nengo.synapses import DiscreteDelay
+
+        dt = 0.001
+        with nengo.Network() as model:
+            stim = nengo.Node(output=lambda t: np.sin(2*np.pi*t))
+            p_stim = nengo.Probe(stim)
+            p_delay = nengo.Probe(stim, synapse=DiscreteDelay(int(0.3 / dt)))
+
+        with nengo.Simulator(model, dt=dt) as sim:
+            sim.run(1.)
+
+        plt.plot(sim.trange(), sim.data[p_stim], label="Stimulus")
+        plt.plot(sim.trange(), sim.data[p_delay], label="Delayed")
+        plt.xlabel("Time (s)")
+        plt.legend()
+
+    .. testoutput::
+       :hide:
+
+       ...
+    """
+
+    steps = IntParam("steps", low=0)
+
+    def __init__(self, steps, **kwargs):
+        self.steps = steps
+        super().__init__(([1], [1] + [0] * steps), analog=False, **kwargs)
+
+
+class LegendreDelay(LinearFilter):
+    r"""A finite-order approximation of a pure delay, using the shifted Legendre basis.
+
+    Implements the transfer function:
+
+    .. math:: H(s) = e^{-\theta s} + \mathcal{O}(s^{2 n})
+
+    where :math:`n` is the order of the system. This results in a pure delay of
+    :math:`\theta` seconds (i.e. :math:`y(t) \approx u(t - \theta)` in the time-domain),
+    for slowly changing inputs.
+
+    Its canonical state-space realization represents the window of history
+    by the shifted Legendre polnomials:
+
+    .. math:: P_i(2 \theta' \theta^{-1} - 1)
+
+    where ``i`` is the zero-based index into the state-vector.
+
+    Parameters
+    ----------
+    theta : ``float``
+        Length of time-delay in seconds.
+    order : ``integer``
+        Order of approximation in the denominator
+        (dimensionality of resulting system).
+
+    See Also
+    --------
+    :class:`.DiscreteDelay`
+
+    Examples
+    --------
+    Delay 15 Hz band-limited white noise by 100 ms using various orders of
+    approximations:
+
+    .. testcode::
+
+        process = nengo.processes.WhiteSignal(10.0, high=15, y0=0)
+        u = process.run_steps(500)
+        t = process.ntrange(len(u))
+
+        plt.plot(t[100:], u[:-100], linestyle="--", lw=4, label="Ideal")
+        for order in list(range(4, 8)):
+            sys = nengo.synapses.LegendreDelay(0.1, order=order)
+            plt.plot(t, sys.filt(u), label="order=%s" % order)
+
+        plt.xlabel("Time (s)")
+        plt.legend()
+    """
+
+    theta = NumberParam("theta", low=0)
+    order = IntParam("order", low=1)
+
+    def __init__(self, theta, order, **kwargs):
+        self.theta = theta
+        self.order = order
+
+        Q = np.arange(order, dtype=np.float64)
+        R = (2 * Q + 1)[:, None] / theta
+        j, i = np.meshgrid(Q, Q)
+
+        A = np.where(i < j, -1, (-1.0) ** (i - j + 1)) * R
+        B = (-1.0) ** Q[:, None] * R
+        C = np.ones((1, order))
+        D = np.zeros((1,))
+
+        super().__init__((A, B, C, D), analog=True, **kwargs)
```

Comment:
Note to selves: following this PR, we can update examples such as https://www.nengo.ai/nengo-dl/examples/lmu.html to obtain the (A, B) matrices by instantiating this class. :)